### PR TITLE
[MRG] Be compatible with newer scikit-learn

### PR DESCRIPTION
**Pull request by wdevazelhes**
Scikit-learn just released their last version 0.21, which does not support python<3.5, so I think we can update the travis script to download the 0.20.3 for python3.4 and python2.7, but still use the newest release for python3.6 (so that we test the behaviour on the new scikit-learn), what do you think ? 

Also, there were a few changes that need to be made to work with scikit-learn's new version, which this PR addresses

### CODE REVIEW

File: metric_learn/_util.py [hunk 1]
```diff
@@ -88,10 +87,6 @@ def check_input(input_data, y=None, preprocessor=None,
     is originally 1D and ``ensure_2d`` is True. Setting to 0 disables
     this check.
 
-  warn_on_dtype : boolean (default=False)
```
Comment (wdevazelhes) [inline metric_learn/_util.py]: scikit-learn 0.21 deprecates warn_on_dtype (see: https://github.com/scikit-learn/scikit-learn/issues/13324). Since we never use it (except to test it works), I suggest we get rid of it

File: test/test_base_metric.py [hunk 2]
```diff
@@ -5,86 +5,108 @@
 from sklearn import clone
 from sklearn.utils.testing import set_random_state
 from test.test_utils import ids_metric_learners, metric_learners
+import string
```
Comment (perimosocordiae) [inline test/test_base_metric.py]: Is this import used?

File: test/test_base_metric.py [hunk 3]
```diff
@@ -5,86 +5,108 @@
...
+
+
+def remove_spaces(s):
+  s = s.replace(' ', '')
+  s = s.replace('\n', '')
+  return s
```
Comment (perimosocordiae) [inline test/test_base_metric.py]: I'd rather use a regex here:

`return re.sub('\s+', ' ', s)`

File: test/test_base_metric.py [hunk 4]
```diff
@@ -5,86 +5,108 @@
...
+    self.assertEqual(remove_spaces(str(metric_learn.Covariance())),
+                     remove_spaces("Covariance(preprocessor=None)"))
 
   def test_lmnn(self):
     self.assertRegexpMatches(
         str(metric_learn.LMNN()),
         r"(python_)?LMNN\(convergence_tol=0.001, k=3, learn_rate=1e-07, "
-        r"max_iter=1000,\n      min_iter=50, preprocessor=None, "
-        r"regularization=0.5, use_pca=True,\n      verbose=False\)")
+        r"max_iter=1000,\n(\ +)min_iter=50, preprocessor=None, "
```
Comment (perimosocordiae) [inline test/test_base_metric.py]: `\s+` is a bit more readable, IMO

File: test/test_base_metric.py [hunk 5]
```diff
@@ -5,86 +5,108 @@
...
-""".strip('\n'))
-    self.assertEqual(str(metric_learn.ITML_Supervised()), """
+""".strip('\n')))
+    self.assertEqual(remove_spaces(str(metric_learn.ITML_Supervised())),
+                     remove_spaces("""
 ITML_Supervised(A0=None, bounds='deprecated', convergence_threshold=0.001,
         gamma=1.0, max_iter=1000, num_constraints=None,
         num_labeled='deprecated', preprocessor=None, verbose=False)
-""".strip('\n'))
+""".strip('\n')))
```
Comment (perimosocordiae) [inline test/test_base_metric.py]: No need to strip newlines if you're also calling `remove_spaces()`, right?

**Comment by terrytangyuan** [comment]
Yes, @perimosocordiae here's the issue: https://github.com/metric-learn/metric-learn/issues/166

File: test/test_base_metric.py (hunk 2 above)
Comment (wdevazelhes) [reply test/test_base_metric.py]: No, indeed

File: test/test_base_metric.py (hunk 3 above)
Comment (wdevazelhes) [reply test/test_base_metric.py]: That's right, it's better
You meant `return re.sub('\s+', '', s)` to remove the spaces right ?

File: test/test_base_metric.py (hunk 4 above)
Comment (wdevazelhes) [reply test/test_base_metric.py]: I agree, it's better, I didn't know `\s`

File: test/test_base_metric.py (hunk 5 above)
Comment (wdevazelhes) [reply test/test_base_metric.py]: That's right indeed

**Comment by wdevazelhes** [comment]
Thanks for the review @perimosocordiae, I just addressed your comments

File: test/test_base_metric.py (hunk 3 above)
Comment (perimosocordiae) [reply test/test_base_metric.py]: Yeah, should be an empty replacement string.

File: test/test_base_metric.py [hunk 6]
```diff
@@ -7,84 +8,103 @@
...
   def test_lmnn(self):
     self.assertRegexpMatches(
-        str(metric_learn.LMNN()),
-        r"(python_)?LMNN\(convergence_tol=0.001, k=3, learn_rate=1e-07, "
-        r"max_iter=1000,\n      min_iter=50, preprocessor=None, "
-        r"regularization=0.5, use_pca=True,\n      verbose=False\)")
+      str(metric_learn.LMNN()),
+      r"(python_)?LMNN\(convergence_tol=0.001, k=3, learn_rate=1e-07, "
+      r"max_iter=1000,(\s+)min_iter=50, preprocessor=None, "
+      r"regularization=0.5, use_pca=True,(\s+)verbose=False\)")
```
Comment (perimosocordiae) [inline test/test_base_metric.py]: Don't need to wrap the whitespace matchers in parens.

Comment (wdevazelhes) [reply test/test_base_metric.py]: That's right, done

**Comment by perimosocordiae** [comment]
+1 to merge when the travis results come back.

**Comment by wdevazelhes** [comment]
The results are green merging
